### PR TITLE
[OptApp] Integration of NLOPT library

### PR DESCRIPTION
**Pull request by RezaNajian**
This PR makes available NLOPT optimizers in OptimizationApplication. Currently, it supports gradient-based optimizers as:
            **"mma": nlopt.LD_MMA,
            "slsqp": nlopt.LD_SLSQP,
            "lbfgs": nlopt.LD_LBFGS,
            "ccsaq": nlopt.LD_CCSAQ,
            "tnewton_precond_restart": nlopt.LD_TNEWTON_PRECOND_RESTART,
            "tnewton_precond": nlopt.LD_TNEWTON_PRECOND,
            "tnewton_restart": nlopt.LD_TNEWTON_RESTART,
            "augmented_lagrangian": nlopt.AUGLAG**

and gradient-free optimizers as 
            **"neldermead": nlopt.LN_NELDERMEAD,
            "evolutionary": nlopt.GN_ESCH**

The interface is implemented at the Python level, and the NLOPT Python library needs to be in the Python path.

A test case is also made which uses MMA for constrained thickness optimization.

### CODE REVIEW

File: applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+    def Finalize(self):
+        self.__objective.Finalize()
+        for constraint in self.__constraints:
+            constraint.Finalize()
+        self.master_control.Finalize()
+
+    def Solve(self):
+        self.nlopt_optimizer.optimize(self.x0)
+        CallOnAll(self._optimization_problem.GetListOfProcesses("output_processes"), Kratos.OutputProcess.PrintOutput)
```
Comment (philbucher) [inline applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py:172]:
```suggestion
        map(Kratos.OutputProcess.PrintOutput, self._optimization_problem.GetListOfProcesses("output_processes")
```

python can already do that

File: applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    def GetDefaultParameters(cls):
+        return Kratos.Parameters("""{
+            "module"            : "KratosMultiphysics.OptimizationApplication.algorithms",
+            "type"              : "PLEASE_PROVIDE_AN_ALGORITHM_CLASS_NAME",
+            "objective"         : {},
+            "constraints"         : [],
+            "controls"          : [],
+            "echo_level"        : 0,
+            "NLOPT_settings"          : {
+                "algorithm_name"      : "MMA",
```
Comment (philbucher) [inline applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py]:
```suggestion
                "algorithm_name"      : "mma",
```

please follow the style guide

File: applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        for constraint_settings in parameters["constraints"]:
+            self.__constraints.append(StandardizedNLOPTConstraint(constraint_settings, self.master_control, self._optimization_problem))
+
+        # nlopt settings
+        NLOPT_settings = parameters["NLOPT_settings"]
+        NLOPT_settings.ValidateAndAssignDefaults(self.GetDefaultParameters()["NLOPT_settings"])
+
+        # nlopt algorithm
+        self.algorithm_name = NLOPT_settings["algorithm_name"].GetString()
+        if(len(self.__constraints)==0):
```
Comment (philbucher) [inline applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py]:
```suggestion
        if self.__constraints:
```

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_constraint.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            return self.__reference_value
+        else:
+            if self.__unbuffered_data.HasValue("initial_value"):
+                self.__reference_value = self.__unbuffered_data["initial_value"]
+                return self.__reference_value
+            else:
+                raise RuntimeError(f"Response value for {self.GetReponse().GetName()} is not calculated yet.")
+
+    def Initialize(self):
+        super().Initialize()
```
Comment (philbucher) [inline applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_constraint.py]:
```suggestion
```

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_constraint.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+    def CalculateStandardizedValueAndGradients(self, control_field: numpy.ndarray, gradient_field: numpy.ndarray, save_value: bool = True) -> float:
+
+        # update the master control
+        update_status = self.GetMasterControl().Update(control_field)
+        master_control_updated = False
+        for is_updated in update_status.values():
+            if is_updated:
+                master_control_updated = True
+                break
```
Comment (philbucher) [inline applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_constraint.py]:
```suggestion
        master_control_updated = any(update_status)
```

might need to play with the syntax, but `any` should work here

File: applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py (hunk 2 above)
Comment (RezaNajian) [reply applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py]: Tnx, Done !

File: applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py (hunk 3 above)
Comment (RezaNajian) [reply applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py]: Done.

File: applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py (hunk 1 above)
Comment (RezaNajian) [reply applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py:172]: if you do not mind, for being consistent with the rest of the application I would keep it

Comment (philbucher) [reply applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py:172]: I dont mind, but oh boy, reimplementing language features :sweat_smile: 

I dont recommend it. I would rather fix the rest of the app

Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py:172]: That would be my fault due to my ignorance... :/

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_constraint.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        ">"
+
+    The reference value for the constraint either can be the "initial_value" or a specified value.
+
+    """
+    def __init__(self, parameters: Kratos.Parameters, master_control: MasterControl, optimization_problem: OptimizationProblem, required_buffer_size: int = 2):
+        default_parameters = Kratos.Parameters("""{
+            "response_name"   : "",
+            "type"            : "",
+            "ref_value": "initial_value"
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_constraint.py]:
```suggestion
            "response_name": "",
            "type"                     : "",
            "ref_value"             : "initial_value"
```

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_constraint.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        self.__unbuffered_data = component_data_view.GetUnBufferedData()
+
+        if parameters["ref_value"].IsDouble():
+            self.__ref_type = "specified_value"
+            self.__reference_value = parameters["ref_value"].GetDouble()
+        elif parameters["ref_value"].IsString() and parameters["ref_value"].GetString() == "initial_value":
+            self.__ref_type = "initial_value"
+            self.__reference_value = None
+        else:
+            raise RuntimeError(f"Provided \"reference_type\" = {self.__ref_type} is not supported for constraint response functions. Followings are supported options: \n\tinitial_value\n\tfloat value")
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_constraint.py]:
```suggestion
        if parameters["ref_value"].IsDouble():
            self.__reference_value = parameters["ref_value"].GetDouble()
        elif parameters["ref_value"].IsString() and parameters["ref_value"].GetString() == "initial_value":
            self.__reference_value = None
        else:
            raise RuntimeError("Provided \"reference_type\" is not supported for constraint response functions. Followings are supported options: \n\tinitial_value\n\tfloat value")
```

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_constraint.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+    def GetStandardizedValue(self, step_index: int = 0) -> float:
+        ref_value = self.GetReferenceValue()
+        standardized_response_value = self.GetValue(step_index) - ref_value
+        standardized_response_value *= self.GetStandardizationFactor()
+        return standardized_response_value
+
+    def GetAbsoluteViolation(self, step_index: int = 0) -> float:
+            ref_value = self.GetReferenceValue()
+            standardized_response_value = self.GetValue() - ref_value
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_constraint.py]:
```suggestion
            standardized_response_value = self.GetValue(step_index) - ref_value
```

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_objective.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+                # log values
+                if save_value:
+                    self.LogValues()
+
+                # print the info
+                DictLogger("Objective info",self.GetInfo())
+
+                # compute standardization factor
+                self.standardization_factor = self.__scaling / abs(self.__unbuffered_data["initial_value"])
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_objective.py]: What if the `initial_value` is zero or close to zero?

File: applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py [hunk 10]
```diff
@@ -205,7 +205,7 @@ def Update(self, new_control_field: ContainerExpressionTypes) -> bool:
         if not IsSameContainerExpression(new_control_field, self.GetEmptyField()):
             raise RuntimeError(f"Updates for the required element container not found for control \"{self.GetName()}\". [ required model part name: {self.model_part.FullName()}, given model part name: {new_control_field.GetModelPart().FullName()} ]")
 
-        if KratosOA.ExpressionUtils.NormL2(self.control_field - new_control_field) > 1e-9:
+        if KratosOA.ExpressionUtils.NormL2(self.control_field - new_control_field) > 1e-15:
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py:208]: May be we can go ahead with sommething like relative change in here?

File: applications/OptimizationApplication/tests/algorithm_tests/nlopt_tests/mma_shell_thickness_opt/optimization_parameters.json [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        "type": "NLOPT_algorithms",
+        "NLOPT_settings"          : {
+                "algorithm_name"      : "mma",
+                "controls_lower_bound" : "0",
+                "controls_upper_bound" : "1",
+                "stopping_criteria"   : {
+                    "maximum_function_evalualtion": 2
+                },
+                "algorithm_specific_settings"   : {
+                "inner_maxeval" : 20}
```
Comment (sunethwarna) [inline applications/OptimizationApplication/tests/algorithm_tests/nlopt_tests/mma_shell_thickness_opt/optimization_parameters.json]:
```suggestion
                "algorithm_specific_settings"   : {
                    "inner_maxeval" : 20
                }
```

OCD thingy ;)

**Review by sunethwarna — COMMENTED** [review]
Nice work @RezaNajian Its cool to see the hardwork we put in last few months pays off. Apart from the minor comments I have mentioned in-place, my major comment is, Could you please reduce the mdpa size. It is 4k+ lines, and whole PR is 5k+ :/

**Comment by RezaNajian** [comment]
@sunethwarna I addressed your points could please accept it ?!

**Review by sunethwarna — DISMISSED** [review]
Thanks @RezaNajian :)

**Comment by RezaNajian** [comment]
@KratosMultiphysics/technical-committee I have been trying to merge this PR but CI/CD on debug mode fails. Could you please guid me to fix the issue ?! Thanks in advance.

**Comment by matekelemen** [comment]
> @KratosMultiphysics/technical-committee @philbucher @sunethwarna I have been trying to merge this PR but CI/CD on debug mode fails. Could you please guid me to fix the issue ?! Thanks in advance.

#11564

**Review by sunethwarna — APPROVED** [review]
Thanks @RezaNajian. You can merge this. The FullDebug versions are for the time being disabled in the CI. (Every developer has to check them in their PCs). I also checked this PR, hence approving :D

**Comment by RezaNajian** [comment]
> Thanks @RezaNajian. You can merge this. The FullDebug versions are for the time being disabled in the CI. (Every developer has to check them in their PCs). I also checked this PR, hence approving :D

Thanks Suneth 🙏

**Pull request merged**